### PR TITLE
Change API to accept names without the path and the lib prefix

### PR DESCRIPTION
**Pull request by martin-olivier**

### CODE REVIEW

File: include/dylib.hpp [hunk 1]
```diff
@@ -31,16 +31,25 @@
 class dylib {
 public:
 #if defined(_WIN32) || defined(_WIN64)
+    static constexpr const char *prefix = "";
```
Comment (eyalroz) [inline include/dylib.hpp:34]: 1. This doesn't look like it should be public.
2. 
Consider:
```
struct {
  static constexpr const char *prefix = "";
  static constexpr const char *suffix = ".dll";
} filename_components

File: include/dylib.hpp [hunk 2]
```diff
@@ -31,16 +31,25 @@
...
     static constexpr const char *extension = ".dll";
```
Comment (eyalroz) [inline include/dylib.hpp:35]: If you want to use "extension", you don't specify the dot. I suggest making it "suffix".

File: include/dylib.hpp [hunk 3]
```diff
@@ -31,16 +31,25 @@
...
+    static constexpr const char *prefix = "lib";
     static constexpr const char *extension = ".dylib";
     using native_handle_type = void *;
 #else
+    static constexpr const char *prefix = "lib";
     static constexpr const char *extension = ".so";
     using native_handle_type = void *;
 #endif
 
+    static constexpr auto os_library_paths = "";
```
Comment (eyalroz) [inline include/dylib.hpp:47]: 1. You only define this once.
2. You say "paths" plural, but it's a single string.
3. The value doesn't seem to make sense.

File: include/dylib.hpp [hunk 4]
```diff
@@ -31,16 +31,25 @@
...
+    static constexpr auto working_directory = "./";
+
+    static constexpr bool add_decorations = true;
```
Comment (eyalroz) [inline include/dylib.hpp:50]: 1. What decorations? Explain with a comment.
2. Why is this public?

File: include/dylib.hpp [hunk 5]
```diff
@@ -92,56 +101,21 @@ class dylib {
...
     /**
      *  Creates a dynamic library instance
      *
-     *  @param path path to the dynamic library to load
-     *  @param ext use dylib::extension to specify the os extension (optional parameter)
+     *  @param path the path where is located the dynamic library you want to load
+     *  @param name the name of the dynamic library to load
+     *  @param decorations add os decorations to the library name
      */
-    explicit dylib(const char *path) {
```
Comment (eyalroz) [inline include/dylib.hpp:103]: Why are we losing the one-parameter constructor? `dylib("foo")`?

File: include/dylib.hpp [hunk 6]
```diff
@@ -220,24 +194,6 @@ class dylib {
...
-     *  This function will be automatically called by the class destructor
-     */
-    void close() noexcept {
-        if (m_handle) {
-            _close(m_handle);
-            m_handle = nullptr;
-        }
-    }
-
 private:
```
Comment (eyalroz) [inline include/dylib.hpp:197]: Why not `protected`?

File: include/dylib.hpp [hunk 7]
```diff
@@ -276,20 +232,47 @@ class dylib {
         return dlerror();
     }
 #endif
+    void open(const std::string &path, const std::string &name, bool decorations) {
+        std::string final_name = name;
+        std::string final_path = path;
+
+        if (decorations)
+            final_name = prefix + name + extension;
```
Comment (eyalroz) [inline include/dylib.hpp:240]: So why did you make a copy of name, before, if you're replacing it?

File: include/dylib.hpp [hunk 8]
```diff
@@ -276,20 +232,47 @@ class dylib {
         return dlerror();
     }
 #endif
+    void open(const std::string &path, const std::string &name, bool decorations) {
```
Comment (eyalroz) [inline include/dylib.hpp:235]: Why is the inner function taking `std::string`'s?

File: include/dylib.hpp [hunk 9]
```diff
@@ -276,20 +232,47 @@ class dylib {
...
+            final_path += '/';
+
+        m_handle = _open((final_path + final_name).c_str());
+
+        if (!m_handle)
+            throw handle_error(get_handle_error(final_path + final_name));
+    }
+
+    void close() noexcept {
+        if (m_handle) {
```
Comment (eyalroz) [inline include/dylib.hpp:252]: How can the handle not be valid at this point?

File: include/dylib.hpp [hunk 10]
```diff
@@ -92,56 +101,21 @@ class dylib {
...
-        open(path);
-    }
-
-    explicit dylib(const std::string &path) {
-        open(path.c_str());
-    }
-
-    dylib(const std::string &path, const char *ext) {
-        open(path, ext);
+    dylib(const std::string &path, const std::string &name, bool decorations = true) {
```
Comment (eyalroz) [inline include/dylib.hpp:111]: 1. Path - to where? Directory path perhaps?
2. Doxygen comment here...

File: include/dylib.hpp [hunk 11]
```diff
@@ -92,56 +101,21 @@ class dylib {
...
+        open(path, name, decorations);
     }
 
     ~dylib() {
         close();
```
Comment (eyalroz) [inline include/dylib.hpp:116]: Drop the close function and inline its contents here. There's no sense in closing before destruction...

File: include/dylib.hpp [hunk 12]
```diff
@@ -276,20 +232,47 @@ class dylib {
...
+        if (path != dylib::os_library_paths && path.find_last_of('/') != path.size() - 1)
+            final_path += '/';
+
+        m_handle = _open((final_path + final_name).c_str());
+
+        if (!m_handle)
+            throw handle_error(get_handle_error(final_path + final_name));
+    }
+
+    void close() noexcept {
```
Comment (eyalroz) [inline include/dylib.hpp:251]: Drop this method, see earlier comment.

File: include/dylib.hpp [hunk 13]
```diff
@@ -276,20 +232,47 @@ class dylib {
...
+        if (m_handle) {
+            _close(m_handle);
+            m_handle = nullptr;
```
Comment (eyalroz) [inline include/dylib.hpp:254]: No need.

File: tests/tests.cpp [hunk 14]
```diff
@@ -31,7 +31,7 @@ TEST(exemple, exemple_test) {
     OSRedirector oss(std::cout);
 
     try {
-        dylib lib("./dynamic_lib", dylib::extension);
+        dylib lib(dylib::working_directory, "dynamic_lib");
```
Comment (eyalroz) [inline tests/tests.cpp:34]: This is not clear enough:

1. `working_directory` does say _which_ directory, and the class-namespace implies it's dylib's working directory, whatever that would be
2. I'm looking for a verb indication that the current working directory is to be used.  But... perhaps there's no decent way to achieve that.

File: tests/tests.cpp [hunk 15]
```diff
@@ -53,7 +53,7 @@ TEST(exemple, exemple_test) {
 
 TEST(ctor, bad_library) {
     try {
-        dylib lib("./null.so");
+        dylib lib(dylib::working_directory, "null");
```
Comment (eyalroz) [inline tests/tests.cpp:56]: A string whose text is null? Confusing. Better say "no_such_library".

And you could also add a testcase for actually passing nullptr if you like.

File: tests/tests.cpp [hunk 16]
```diff
@@ -62,44 +62,13 @@ TEST(ctor, bad_library) {
...
-    try {
-        dylib lib("./dynamic_lib", dylib::extension);
-        lib.close();
-        lib.get_function<double(double, double)>("adder");
-        EXPECT_EQ(true, false);
-    }
-    catch (const dylib::handle_error &) {
-        EXPECT_EQ(true, true);
-    }
+    dylib libA(dylib::working_directory, "dynamic_lib");
```
Comment (eyalroz) [inline tests/tests.cpp:65]: I don't like all of these "working_directory" uses, nor this constant itself... 

1. It's not common to search for libraries in the current directory, you might not even want to accommodate that especially.
2. If you must - have a named constructor idiom for it, or a proxy object etc.

**Review by eyalroz — CHANGES_REQUESTED** [review]

File: tests/tests.cpp (hunk 16 above)
Comment (martin-olivier) [reply tests/tests.cpp]: Since i removed `os_library_path` and `current_directory`, i replaced that with `"./"`

File: include/dylib.hpp (hunk 6 above)
Comment (eyalroz) [reply include/dylib.hpp:220]: You never know why people might inherit form your class.

Would you like an example: A child class which logs the symbols being used from the DLL and the time they were obtained, to examine the pattern of library usage, offline.

If there's no benefit to you in `private`izing members, make them protected by default.

File: include/dylib.hpp (hunk 13 above)
Comment (eyalroz) [reply include/dylib.hpp:277]: And if you adopt my earlier comments, the handle is never null.

File: include/dylib.hpp (hunk 11 above)
Comment (martin-olivier) [reply include/dylib.hpp:139]: Done

File: include/dylib.hpp (hunk 12 above)
Comment (martin-olivier) [reply include/dylib.hpp:274]: Done

File: include/dylib.hpp (hunk 13 above)
Comment (martin-olivier) [reply include/dylib.hpp:277]: Done

File: include/dylib.hpp [hunk 17]
```diff
@@ -11,6 +11,7 @@
 
 #pragma once
 
+#include <iostream>
```
Comment (eyalroz) [inline include/dylib.hpp:14]: Did you intend to add iostream here? Not sure it makes sense.

Comment (martin-olivier) [reply include/dylib.hpp:14]: I forgot to remove it after some debug.
Done

**Comment by martin-olivier** [comment]
I've made your request changes.
Don't hesitate to resolve conversations if you are fine with my changes

**Pull request merged**